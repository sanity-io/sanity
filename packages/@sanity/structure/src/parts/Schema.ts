import getDefaultModule from './getDefaultModule'
import {Ordering} from '../Sort'

interface Schema {
  name: string
  get(typeName: string): SchemaType | undefined
  getTypeNames(): string[]
}

interface SchemaField {
  name: string
  type: SchemaType
}

interface PreviewFields {
  media?: string
}

interface PreviewPreparer {
  (selection: {}): PreviewFields
}

export interface SchemaType {
  name: string
  title?: string
  icon?: Function
  type?: SchemaType
  to?: SchemaField[]
  fields?: SchemaField[]
  orderings?: Ordering[]
  initialValue?: Function | {[key: string]: any}
  preview?: {
    select?: PreviewFields
    prepare?: PreviewPreparer
  }
}

// We are lazy-loading the part to work around typescript trying to resolve it
const getDefaultSchema = (): Schema => {
  const schema: Schema = getDefaultModule(require('part:@sanity/base/schema'))
  return schema
}

export {Schema, getDefaultSchema}
