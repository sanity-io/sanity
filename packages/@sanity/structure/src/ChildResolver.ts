import {CollectionBuilder, Collection, SerializePath, SerializeOptions} from './StructureNodes'

export interface ChildResolverOptions {
  index: number
  parent: Collection
  parentPath?: SerializePath
  parameters?: {[key: string]: any}
  serializeOptions?: SerializeOptions
}

export type ItemChild = CollectionBuilder | Collection | Function | undefined

interface ChildObservable {
  subscribe: (child: ItemChild | Promise<ItemChild>) => {}
}

export interface ChildResolver {
  (itemId: string, options: ChildResolverOptions):
    | ItemChild
    | Promise<ItemChild>
    | ChildObservable
    | undefined
}
