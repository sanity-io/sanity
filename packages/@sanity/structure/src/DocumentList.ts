import {getParameterlessTemplatesBySchemaType} from '@sanity/initial-value-templates'
import {SchemaType} from './parts/Schema'
import {client} from './parts/Client'
import {SortItem} from './Sort'
import {EditorBuilder} from './Editor'
import {SerializeError, HELP_URL} from './SerializeError'
import {SerializeOptions, Child} from './StructureNodes'
import {ChildResolver, ChildResolverOptions, ItemChild} from './ChildResolver'
import {InitialValueTemplateItem} from './InitialValueTemplateItem'
import {
  GenericListBuilder,
  BuildableGenericList,
  GenericList,
  GenericListInput
} from './GenericList'

const resolveTypeForDocument = (id: string): Promise<string | undefined> => {
  const query = '*[_id in [$documentId, $draftId]]._type'
  const documentId = id.replace(/^drafts\./, '')
  const draftId = `drafts.${documentId}`
  return client.fetch(query, {documentId, draftId}).then(types => types[0])
}

const validateFilter = (spec: PartialDocumentList, options: SerializeOptions) => {
  const filter = spec.options!.filter.trim()

  if (['*', '{'].includes(filter[0])) {
    throw new SerializeError(
      `\`filter\` cannot start with \`${filter[0]}\` - looks like you are providing a query, not a filter`,
      options.path,
      spec.id,
      spec.title
    ).withHelpUrl(HELP_URL.QUERY_PROVIDED_FOR_FILTER)
  }

  return filter
}

const resolveEditorChildForItem: ChildResolver = (
  itemId: string,
  options: ChildResolverOptions
): ItemChild | Promise<ItemChild> | undefined => {
  const parentItem = options.parent as DocumentList
  const schemaType = parentItem.schemaTypeName || resolveTypeForDocument(itemId)
  return Promise.resolve(schemaType).then(type =>
    new EditorBuilder()
      .id('editor')
      .documentId(itemId)
      .schemaType(type || '')
  )
}

export interface PartialDocumentList extends BuildableGenericList {
  options?: DocumentListOptions
  schemaTypeName?: string
}

export interface DocumentListInput extends GenericListInput {
  options: DocumentListOptions
}

export interface DocumentList extends GenericList {
  options: DocumentListOptions
  child: Child
  schemaTypeName?: string
}

interface DocumentListOptions {
  filter: string
  params?: {[key: string]: any}
  defaultOrdering?: SortItem[]
}

export class DocumentListBuilder extends GenericListBuilder<
  PartialDocumentList,
  DocumentListBuilder
> {
  protected spec: PartialDocumentList

  constructor(spec?: DocumentListInput) {
    super()
    this.spec = spec ? spec : {}
  }

  filter(filter: string): DocumentListBuilder {
    return this.clone({options: {...(this.spec.options || {}), filter}})
  }

  getFilter() {
    return this.spec.options && this.spec.options.filter
  }

  schemaType(type: SchemaType | string): DocumentListBuilder {
    const schemaTypeName = typeof type === 'string' ? type : type.name
    return this.clone({schemaTypeName})
  }

  getSchemaType() {
    return this.spec.schemaTypeName
  }

  params(params: {}): DocumentListBuilder {
    return this.clone({options: {...(this.spec.options || {filter: ''}), params}})
  }

  getParams() {
    return this.spec.options && this.spec.options.params
  }

  defaultOrdering(ordering: SortItem[]): DocumentListBuilder {
    if (!Array.isArray(ordering)) {
      throw new Error('`defaultOrdering` must be an array of order clauses')
    }

    return this.clone({
      options: {...(this.spec.options || {filter: ''}), defaultOrdering: ordering}
    })
  }

  getDefaultOrdering() {
    return this.spec.options && this.spec.options.defaultOrdering
  }

  serialize(options: SerializeOptions = {path: []}): DocumentList {
    if (typeof this.spec.id !== 'string' || !this.spec.id) {
      throw new SerializeError(
        '`id` is required for document lists',
        options.path,
        options.index,
        this.spec.title
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!this.spec.options || !this.spec.options.filter) {
      throw new SerializeError(
        '`filter` is required for document lists',
        options.path,
        this.spec.id,
        this.spec.title
      ).withHelpUrl(HELP_URL.FILTER_REQUIRED)
    }

    return {
      ...super.serialize(options),
      type: 'documentList',
      schemaTypeName: this.spec.schemaTypeName,
      child: this.spec.child || resolveEditorChildForItem,
      options: {
        ...this.spec.options,
        filter: validateFilter(this.spec, options)
      }
    }
  }

  clone(withSpec?: PartialDocumentList): DocumentListBuilder {
    const builder = new DocumentListBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}

    if (!builder.spec.initialValueTemplates) {
      builder.spec.initialValueTemplates = inferInitialValueTemplates(builder.spec)
    }

    return builder
  }

  getSpec() {
    return this.spec
  }
}

function inferInitialValueTemplates(
  spec: PartialDocumentList
): InitialValueTemplateItem[] | undefined {
  const {schemaTypeName, options} = spec
  const {filter, params} = options || {filter: '', params: {}}
  const typeName = schemaTypeName || getTypeNameFromSingleTypeFilter(filter, params)

  if (!typeName) {
    return undefined
  }

  return getParameterlessTemplatesBySchemaType(typeName).map(tpl => ({
    type: 'initialValueTemplateItem',
    id: tpl.id,
    templateId: tpl.id
  }))
}

export function getTypeNameFromSingleTypeFilter(
  filter: string,
  params: {[key: string]: any} = {}
): string | null {
  const pattern = /\b_type\s*==\s*(['"].*?['"]|\$.*?(?:\s|$))|\B(['"].*?['"]|\$.*?(?:\s|$))\s*==\s*_type\b/
  const matches = filter.match(pattern)
  if (!matches) {
    return null
  }

  const match = (matches[1] || matches[2]).trim().replace(/^["']|["']$/g, '')
  const typeName = match[0] === '$' ? params[match.slice(1)] : match
  return typeName || null
}
