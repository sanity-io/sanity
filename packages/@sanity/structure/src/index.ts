import {uniqueId} from 'lodash'
import {ListBuilder, ListInput} from './List'
import {
  getDocumentTypeListItems,
  getDocumentTypeListItem,
  getDocumentTypeList
} from './documentTypeListItems'
import {
  MenuItemBuilder,
  MenuItem,
  getOrderingMenuItemsForSchemaType,
  getOrderingMenuItem
} from './MenuItem'
import {ListItemBuilder, ListItemInput} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {EditorBuilder, editorWithInitialValueTemplate} from './Editor'
import {EditorNode, Divider} from './StructureNodes'
import {SerializeError} from './SerializeError'
import {ComponentInput, ComponentBuilder} from './Component'
import {DocumentListItemBuilder, DocumentListItemInput} from './DocumentListItem'
import {Ordering} from './Sort'
import {SchemaType} from './parts/Schema'
import {
  InitialValueTemplateItemBuilder,
  defaultInitialValueTemplateItems,
  menuItemsFromInitialValueTemplateItems
} from './InitialValueTemplateItem'

const StructureBuilder = {
  defaults: getDefaultStructure,
  documentTypeList: getDocumentTypeList,
  documentTypeListItem: getDocumentTypeListItem,
  documentTypeListItems: getDocumentTypeListItems,

  list: (spec?: ListInput) => new ListBuilder(spec),
  listItem: (spec?: ListItemInput) => new ListItemBuilder(spec),

  menuItem: (spec?: MenuItem) => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup) => new MenuItemGroupBuilder(spec),
  menuItemsFromInitialValueTemplateItems,

  documentList: (spec?: DocumentListInput) => new DocumentListBuilder(spec),
  documentListItem: (spec?: DocumentListItemInput) => new DocumentListItemBuilder(spec),

  orderingMenuItem: (ordering: Ordering) => getOrderingMenuItem(ordering),
  orderingMenuItemsForType: (type: string) => getOrderingMenuItemsForSchemaType(type),

  editor: (spec?: EditorNode) => new EditorBuilder(spec),
  editorWithInitialValueTemplate,

  defaultInitialValueTemplateItems,
  initialValueTemplateItem: (
    templateId: string,
    parameters?: {[key: string]: any}
  ): InitialValueTemplateItemBuilder =>
    new InitialValueTemplateItemBuilder({
      id: templateId,
      parameters,
      templateId
    }),

  component: (spec?: ComponentInput | Function) => {
    return typeof spec === 'function'
      ? new ComponentBuilder().component(spec)
      : new ComponentBuilder(spec)
  },

  divider: (): Divider => ({id: uniqueId('__divider__'), type: 'divider'})
}

function hasIcon(schemaType?: SchemaType | string): boolean {
  if (!schemaType || typeof schemaType === 'string') {
    return false
  }

  return Boolean(schemaType.icon)
}

function getDefaultStructure(): ListBuilder {
  const items = getDocumentTypeListItems()
  return new ListBuilder()
    .id('__root__')
    .title('Content')
    .items(items)
    .showIcons(items.some(item => hasIcon(item.getSchemaType())))
}

export {StructureBuilder, SerializeError}
