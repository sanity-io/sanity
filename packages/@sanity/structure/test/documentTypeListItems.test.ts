import {StructureBuilder as S} from '../src'
import {getDefaultSchema} from '../src/parts/Schema'
import {DocumentList} from '../src/DocumentList'
import serializeStructure from './util/serializeStructure'

const nope = () => 'NOPE'
const editor = {
  id: 'editor',
  options: {
    type: 'author',
    id: 'grrm'
  }
}
const defaultSchema = getDefaultSchema()

test('generates correct document type list items from global schema', () => {
  expect(S.documentTypeListItems()).toMatchSnapshot()
})

test('generates correct document type list items from given schema', () => {
  expect(S.documentTypeListItems(defaultSchema)).toMatchSnapshot()
})

test('generates correct document type list item for specific type', () => {
  expect(S.documentTypeListItem('author')).toMatchSnapshot()
  expect(S.documentTypeListItem('author', defaultSchema)).toMatchSnapshot()
})

test('generated canHandleIntent responds to edit/create on document type', () => {
  const listItems = S.documentTypeListItems().map(item => item.serialize())
  expect(listItems).toHaveLength(2)
  expect(listItems[0]).toMatchObject({id: 'author', title: 'Author'})
  expect(listItems[0]).toHaveProperty('child')
  expect(listItems[0].child).toHaveProperty('canHandleIntent')

  const child = listItems[0].child as DocumentList
  const ctx = {pane: child}
  expect((child.canHandleIntent || nope)('create', {type: 'book'}, ctx)).toBe(false)
  expect((child.canHandleIntent || nope)('create', {type: 'author'}, ctx)).toBe(true)
  expect((child.canHandleIntent || nope)('edit', {id: 'wow'}, ctx)).toBe(false)
})

test('generated document panes responds with correct editor child', done => {
  const listItems = S.documentTypeListItems().map(item => item.serialize())
  expect(listItems).toHaveLength(2)
  expect(listItems[0]).toMatchObject({id: 'author', title: 'Author'})
  expect(listItems[0]).toHaveProperty('child')

  const listItem = listItems[0].child as DocumentList
  const context = {parent: listItem, index: 0}
  serializeStructure(listItem.child, context, ['grrm', context]).subscribe(itemChild => {
    expect(itemChild).toMatchObject(editor)

    done()
  })
})

test('manually assigned canHandleIntent should not be overriden', () => {
  const alwaysFalse = () => false
  const list = S.documentTypeList('author')
  const modified = list.canHandleIntent(alwaysFalse)
  const ctx = {pane: list.serialize()}

  // Test default handler
  const defHandler = list.getCanHandleIntent() || nope
  expect(defHandler('create', {type: 'author'}, ctx)).toBe(true)

  // Test modified handler
  const modHandler = modified.getCanHandleIntent() || nope
  expect(modHandler('create', {type: 'author'}, ctx)).toBe(false)

  // Modifying child for default list _SHOULD_ reset canHandleIntent
  const diffChild = list.child(() => editor)
  const diffHandler = diffChild.getCanHandleIntent()
  expect(diffHandler).toBeUndefined()

  // Modifying child for list with custom intent handler should _NOT_ reset canHandleIntent
  const customChild = modified.child(() => editor)
  const customHandler = customChild.getCanHandleIntent()
  expect(customHandler).toEqual(alwaysFalse)
})
