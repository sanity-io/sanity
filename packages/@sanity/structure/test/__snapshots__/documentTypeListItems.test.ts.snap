// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates correct document type list item for specific type 1`] = `
ListItemBuilder {
  "spec": Object {
    "child": DocumentTypeListBuilder {
      "spec": Object {
        "canHandleIntent": [Function],
        "child": [Function],
        "displayOptions": Object {
          "showIcons": true,
        },
        "id": "author",
        "initialValueTemplates": Array [],
        "menuItemGroups": Array [
          Object {
            "id": "sorting",
            "title": "Sort",
          },
          Object {
            "id": "layout",
            "title": "Layout",
          },
          Object {
            "id": "actions",
            "title": "Actions",
          },
        ],
        "menuItems": Array [
          MenuItemBuilder {
            "spec": Object {
              "action": "setSortOrder",
              "group": "sorting",
              "icon": [Function],
              "params": Object {
                "by": Array [
                  Object {
                    "direction": "desc",
                    "field": "_updatedAt",
                  },
                ],
                "extendedProjection": "",
              },
              "title": "Sort by Last edited",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setSortOrder",
              "group": "sorting",
              "icon": [Function],
              "params": Object {
                "by": Array [
                  Object {
                    "direction": "desc",
                    "field": "_createdAt",
                  },
                ],
                "extendedProjection": "",
              },
              "title": "Sort by Created",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setLayout",
              "group": "layout",
              "icon": [Function],
              "params": Object {
                "layout": "default",
              },
              "title": "List",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setLayout",
              "group": "layout",
              "icon": [Function],
              "params": Object {
                "layout": "detail",
              },
              "title": "Details",
            },
          },
        ],
        "options": Object {
          "defaultOrdering": Array [
            Object {
              "direction": "desc",
              "field": "_updatedAt",
            },
          ],
          "filter": "_type == $type",
          "params": Object {
            "type": "author",
          },
        },
        "schemaTypeName": "author",
        "title": "Author",
      },
    },
    "id": "author",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "name",
          "title": "Name",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "name",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "image",
          "options": Object {
            "hotspot": true,
          },
          "title": "Image",
          "type": "image",
        },
      ],
      "name": "author",
      "preview": Object {
        "select": Object {
          "media": "image",
          "title": "name",
        },
      },
      "title": "Author",
      "type": "document",
    },
    "title": "Author",
  },
}
`;

exports[`generates correct document type list item for specific type 2`] = `
ListItemBuilder {
  "spec": Object {
    "child": DocumentTypeListBuilder {
      "spec": Object {
        "canHandleIntent": [Function],
        "child": [Function],
        "displayOptions": Object {
          "showIcons": true,
        },
        "id": "author",
        "initialValueTemplates": Array [],
        "menuItemGroups": Array [
          Object {
            "id": "sorting",
            "title": "Sort",
          },
          Object {
            "id": "layout",
            "title": "Layout",
          },
          Object {
            "id": "actions",
            "title": "Actions",
          },
        ],
        "menuItems": Array [
          MenuItemBuilder {
            "spec": Object {
              "action": "setSortOrder",
              "group": "sorting",
              "icon": [Function],
              "params": Object {
                "by": Array [
                  Object {
                    "direction": "desc",
                    "field": "_updatedAt",
                  },
                ],
                "extendedProjection": "",
              },
              "title": "Sort by Last edited",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setSortOrder",
              "group": "sorting",
              "icon": [Function],
              "params": Object {
                "by": Array [
                  Object {
                    "direction": "desc",
                    "field": "_createdAt",
                  },
                ],
                "extendedProjection": "",
              },
              "title": "Sort by Created",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setLayout",
              "group": "layout",
              "icon": [Function],
              "params": Object {
                "layout": "default",
              },
              "title": "List",
            },
          },
          MenuItemBuilder {
            "spec": Object {
              "action": "setLayout",
              "group": "layout",
              "icon": [Function],
              "params": Object {
                "layout": "detail",
              },
              "title": "Details",
            },
          },
        ],
        "options": Object {
          "defaultOrdering": Array [
            Object {
              "direction": "desc",
              "field": "_updatedAt",
            },
          ],
          "filter": "_type == $type",
          "params": Object {
            "type": "author",
          },
        },
        "schemaTypeName": "author",
        "title": "Author",
      },
    },
    "id": "author",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "name",
          "title": "Name",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "name",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "image",
          "options": Object {
            "hotspot": true,
          },
          "title": "Image",
          "type": "image",
        },
      ],
      "name": "author",
      "preview": Object {
        "select": Object {
          "media": "image",
          "title": "name",
        },
      },
      "title": "Author",
      "type": "document",
    },
    "title": "Author",
  },
}
`;

exports[`generates correct document type list items from given schema 1`] = `
Array [
  ListItemBuilder {
    "spec": Object {
      "child": DocumentTypeListBuilder {
        "spec": Object {
          "canHandleIntent": [Function],
          "child": [Function],
          "displayOptions": Object {
            "showIcons": true,
          },
          "id": "author",
          "initialValueTemplates": Array [],
          "menuItemGroups": Array [
            Object {
              "id": "sorting",
              "title": "Sort",
            },
            Object {
              "id": "layout",
              "title": "Layout",
            },
            Object {
              "id": "actions",
              "title": "Actions",
            },
          ],
          "menuItems": Array [
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_updatedAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Last edited",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_createdAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Created",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "default",
                },
                "title": "List",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "detail",
                },
                "title": "Details",
              },
            },
          ],
          "options": Object {
            "defaultOrdering": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
            "filter": "_type == $type",
            "params": Object {
              "type": "author",
            },
          },
          "schemaTypeName": "author",
          "title": "Author",
        },
      },
      "id": "author",
      "schemaType": Object {
        "fields": Array [
          Object {
            "name": "name",
            "title": "Name",
            "type": "string",
          },
          Object {
            "name": "slug",
            "options": Object {
              "maxLength": 96,
              "source": "name",
            },
            "title": "Slug",
            "type": "slug",
          },
          Object {
            "name": "image",
            "options": Object {
              "hotspot": true,
            },
            "title": "Image",
            "type": "image",
          },
        ],
        "name": "author",
        "preview": Object {
          "select": Object {
            "media": "image",
            "title": "name",
          },
        },
        "title": "Author",
        "type": "document",
      },
      "title": "Author",
    },
  },
  ListItemBuilder {
    "spec": Object {
      "child": DocumentTypeListBuilder {
        "spec": Object {
          "canHandleIntent": [Function],
          "child": [Function],
          "displayOptions": Object {
            "showIcons": true,
          },
          "id": "post",
          "initialValueTemplates": Array [],
          "menuItemGroups": Array [
            Object {
              "id": "sorting",
              "title": "Sort",
            },
            Object {
              "id": "layout",
              "title": "Layout",
            },
            Object {
              "id": "actions",
              "title": "Actions",
            },
          ],
          "menuItems": Array [
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "asc",
                      "field": "title",
                    },
                    Object {
                      "direction": "asc",
                      "field": "publicationYear",
                    },
                  ],
                  "extendedProjection": "title, publicationYear",
                },
                "title": "Sort by Title",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "asc",
                      "field": "translations.se",
                    },
                    Object {
                      "direction": "asc",
                      "field": "title",
                    },
                  ],
                  "extendedProjection": "translations, title",
                },
                "title": "Sort by Swedish title",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_updatedAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Last edited",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_createdAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Created",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "default",
                },
                "title": "List",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "detail",
                },
                "title": "Details",
              },
            },
          ],
          "options": Object {
            "defaultOrdering": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
            "filter": "_type == $type",
            "params": Object {
              "type": "post",
            },
          },
          "schemaTypeName": "post",
          "title": "Post",
        },
      },
      "id": "post",
      "schemaType": Object {
        "fields": Array [
          Object {
            "name": "title",
            "title": "Title",
            "type": "string",
          },
          Object {
            "name": "slug",
            "options": Object {
              "maxLength": 96,
              "source": "title",
            },
            "title": "Slug",
            "type": "slug",
          },
          Object {
            "name": "author",
            "title": "Author",
            "to": Object {
              "type": "author",
            },
            "type": "reference",
          },
          Object {
            "name": "mainImage",
            "options": Object {
              "hotspot": true,
            },
            "title": "Main image",
            "type": "image",
          },
          Object {
            "name": "publishedAt",
            "title": "Published at",
            "type": "datetime",
          },
          Object {
            "name": "publicationYear",
            "title": "Publication year",
            "type": "string",
          },
          Object {
            "fields": Array [
              Object {
                "name": "se",
                "title": "Swedish",
                "type": "string",
              },
              Object {
                "name": "no",
                "title": "Norwegian",
                "type": "string",
              },
            ],
            "name": "translations",
            "title": "Translations",
            "type": "object",
          },
          Object {
            "name": "body",
            "title": "Body",
            "type": "text",
          },
        ],
        "icon": [Function],
        "initialValue": Object {
          "slug": Object {
            "_type": "slug",
            "current": "default-slug",
          },
        },
        "name": "post",
        "orderings": Array [
          Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "title",
              },
              Object {
                "direction": "asc",
                "field": "publicationYear",
              },
            ],
            "name": "title",
            "title": "Title",
          },
          Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "translations.se",
              },
              Object {
                "direction": "asc",
                "field": "title",
              },
            ],
            "name": "swedishTitle",
            "title": "Swedish title",
          },
        ],
        "preview": Object {
          "select": Object {
            "author": "author.name",
            "media": "mainImage",
            "title": "title",
          },
        },
        "title": "Post",
        "type": "document",
      },
      "title": "Post",
    },
  },
]
`;

exports[`generates correct document type list items from global schema 1`] = `
Array [
  ListItemBuilder {
    "spec": Object {
      "child": DocumentTypeListBuilder {
        "spec": Object {
          "canHandleIntent": [Function],
          "child": [Function],
          "displayOptions": Object {
            "showIcons": true,
          },
          "id": "author",
          "initialValueTemplates": Array [],
          "menuItemGroups": Array [
            Object {
              "id": "sorting",
              "title": "Sort",
            },
            Object {
              "id": "layout",
              "title": "Layout",
            },
            Object {
              "id": "actions",
              "title": "Actions",
            },
          ],
          "menuItems": Array [
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_updatedAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Last edited",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_createdAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Created",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "default",
                },
                "title": "List",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "detail",
                },
                "title": "Details",
              },
            },
          ],
          "options": Object {
            "defaultOrdering": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
            "filter": "_type == $type",
            "params": Object {
              "type": "author",
            },
          },
          "schemaTypeName": "author",
          "title": "Author",
        },
      },
      "id": "author",
      "schemaType": Object {
        "fields": Array [
          Object {
            "name": "name",
            "title": "Name",
            "type": "string",
          },
          Object {
            "name": "slug",
            "options": Object {
              "maxLength": 96,
              "source": "name",
            },
            "title": "Slug",
            "type": "slug",
          },
          Object {
            "name": "image",
            "options": Object {
              "hotspot": true,
            },
            "title": "Image",
            "type": "image",
          },
        ],
        "name": "author",
        "preview": Object {
          "select": Object {
            "media": "image",
            "title": "name",
          },
        },
        "title": "Author",
        "type": "document",
      },
      "title": "Author",
    },
  },
  ListItemBuilder {
    "spec": Object {
      "child": DocumentTypeListBuilder {
        "spec": Object {
          "canHandleIntent": [Function],
          "child": [Function],
          "displayOptions": Object {
            "showIcons": true,
          },
          "id": "post",
          "initialValueTemplates": Array [],
          "menuItemGroups": Array [
            Object {
              "id": "sorting",
              "title": "Sort",
            },
            Object {
              "id": "layout",
              "title": "Layout",
            },
            Object {
              "id": "actions",
              "title": "Actions",
            },
          ],
          "menuItems": Array [
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "asc",
                      "field": "title",
                    },
                    Object {
                      "direction": "asc",
                      "field": "publicationYear",
                    },
                  ],
                  "extendedProjection": "title, publicationYear",
                },
                "title": "Sort by Title",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "asc",
                      "field": "translations.se",
                    },
                    Object {
                      "direction": "asc",
                      "field": "title",
                    },
                  ],
                  "extendedProjection": "translations, title",
                },
                "title": "Sort by Swedish title",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_updatedAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Last edited",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setSortOrder",
                "group": "sorting",
                "icon": [Function],
                "params": Object {
                  "by": Array [
                    Object {
                      "direction": "desc",
                      "field": "_createdAt",
                    },
                  ],
                  "extendedProjection": "",
                },
                "title": "Sort by Created",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "default",
                },
                "title": "List",
              },
            },
            MenuItemBuilder {
              "spec": Object {
                "action": "setLayout",
                "group": "layout",
                "icon": [Function],
                "params": Object {
                  "layout": "detail",
                },
                "title": "Details",
              },
            },
          ],
          "options": Object {
            "defaultOrdering": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
            "filter": "_type == $type",
            "params": Object {
              "type": "post",
            },
          },
          "schemaTypeName": "post",
          "title": "Post",
        },
      },
      "id": "post",
      "schemaType": Object {
        "fields": Array [
          Object {
            "name": "title",
            "title": "Title",
            "type": "string",
          },
          Object {
            "name": "slug",
            "options": Object {
              "maxLength": 96,
              "source": "title",
            },
            "title": "Slug",
            "type": "slug",
          },
          Object {
            "name": "author",
            "title": "Author",
            "to": Object {
              "type": "author",
            },
            "type": "reference",
          },
          Object {
            "name": "mainImage",
            "options": Object {
              "hotspot": true,
            },
            "title": "Main image",
            "type": "image",
          },
          Object {
            "name": "publishedAt",
            "title": "Published at",
            "type": "datetime",
          },
          Object {
            "name": "publicationYear",
            "title": "Publication year",
            "type": "string",
          },
          Object {
            "fields": Array [
              Object {
                "name": "se",
                "title": "Swedish",
                "type": "string",
              },
              Object {
                "name": "no",
                "title": "Norwegian",
                "type": "string",
              },
            ],
            "name": "translations",
            "title": "Translations",
            "type": "object",
          },
          Object {
            "name": "body",
            "title": "Body",
            "type": "text",
          },
        ],
        "icon": [Function],
        "initialValue": Object {
          "slug": Object {
            "_type": "slug",
            "current": "default-slug",
          },
        },
        "name": "post",
        "orderings": Array [
          Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "title",
              },
              Object {
                "direction": "asc",
                "field": "publicationYear",
              },
            ],
            "name": "title",
            "title": "Title",
          },
          Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "translations.se",
              },
              Object {
                "direction": "asc",
                "field": "title",
              },
            ],
            "name": "swedishTitle",
            "title": "Swedish title",
          },
        ],
        "preview": Object {
          "select": Object {
            "author": "author.name",
            "media": "mainImage",
            "title": "title",
          },
        },
        "title": "Post",
        "type": "document",
      },
      "title": "Post",
    },
  },
]
`;
